infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -81,19 +90,74 @@
             .getSerializedSize();
     private final boolean noVerify;
     public final String logDir;
-    private Map<String, SegmentHandle> writeChannels;
+    private LoadingCache<Pair<UUID, Long>, SegmentHandle> segmentsCache;
     private Set<FileChannel> channelsToSync;
     private MultiReadWriteLock segmentLocks = new MultiReadWriteLock();
 
     public StreamLogFiles(String logDir, boolean noVerify) {
+        this(logDir, noVerify, DEFAULT_SEGMENT_CACHE_SIZE);
+    }
+
+    public StreamLogFiles(String logDir, boolean noVerify, long segmentCacheSize) {
         this.logDir = logDir;
-        writeChannels = new ConcurrentHashMap();
+        segmentsCache = getCache(segmentCacheSize);
         channelsToSync = new HashSet<>();
         this.noVerify = noVerify;
 
         verifyLogs();
     }
 
+    @SuppressWarnings("unchecked")
+    private LoadingCache<Pair<UUID, Long>, SegmentHandle> getCache(long numSegments) {
+        return CacheBuilder.newBuilder()
+                .maximumWeight(numSegments)
+                .weigher((Long, SegmentHandle) -> {
+                            return 1;
+                        }
+                )
+                .removalListener(new RemovalListener() {
+                    @Override
+                    public void onRemoval(RemovalNotification notification) {
+                        // Would this cause NPE for other objects holding this segment's reference?
```

Comment:
Theoretically this can happen when you have an object that holds a reference to a write channel, but the segment that corresponds to that channel has been evicted from the cache. Writing to that channel will  cause an NPE because on eviction the channel is closed and set to null.

So, this can happen on a trim, append or ready. Since those operations call getSegmentHandleForAddress before accessing the channels, the NPE can happen only if the corresponding segment is evicted by the time those operations try to use that channel. For that segment to be evicted within that small time frame, a large number of read/write requests must occur, which is extremely unlikely given the large size of the cache and our access pattern.

For example, when writing address 0 (assume the cache is empty), for an NPE to happen 10,000 random read/write(each segment resulting in loading a new segment into the cache) request must happen after getSegmentHandleForAddress(addr 0) and before channel.write(addr 0). 

The possibility of this happening can be reduced further at the expense of increasing the size of the cache, which adds a small reasonably memory overhead.

I think the probability of it occurring is extremely small that it can be ignored.